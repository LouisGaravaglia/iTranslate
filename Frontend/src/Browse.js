import React,  {useState, useRef, useEffect} from 'react';
import './App.css';
//COMPONENT IMPORTS
import FlashMessage from "./FlashMessage";
import BrowseByArtists from "./BrowseByArtists";
import BrowseByGenre from "./BrowseByGenre";
import BrowseByDanceability from "./BrowseByDanceability";
//REDUX IMPORTS
import {useDispatch, useSelector} from "react-redux";
import {resetLanguageError, resetTranslationError, resetLyricsError} from "./actionCreators/handleErrorsCreator";
import {resetStore} from "./actionCreators/resetStoreCreator";
import {getAllArtists} from "./actionCreators/BrowseRoute/Artists/getAllArtistsCreator";
import {getGenres} from "./actionCreators/BrowseRoute/Genre/getGenresCreator";


function Browse() {
  //REACT STATE
  const [category, setCategory] = useState("");
  //REDUX STORE
  const dispatch = useDispatch();
  const languageError = useSelector(store => store.errors.languageError);
  const translationError = useSelector(store => store.errors.translationError);
  const lyricsError = useSelector(store => store.errors.lyricsError);
  //STATE FOR FLASH MESSAGES
  const [searchFlashMessage, setSearchFlashMessage] = useState(false);
  const [noLyricsFlashMessage, setNoLyricsFlashMessage] = useState(false);
  const [languageNotFoundFlashMessage, setLanguageNotFoundFlashMessage] = useState(false);
  const [translationErrorFlashMessage, setTranslationErrorFlashMessage] = useState(false);
  const [noAlbumsFlashMessage, setNoAlbumsFlashMessage] = useState(false);
  //REFS FOR PAGE TRAVERSAL
  const artistResultsRef = useRef();
  const genreResultsRef = useRef();
  const danceabilityResultsRef = useRef();

////////////////////////////////////////////////////  USE EFFECTS  ////////////////////////////////////////////////////

  useEffect(() => {

    async function getSeedData() {
      //GET ALL ARTISTS IN DB AND STORE THEM FOR THE BROWSE BY ARTISTS COMPONENT
      dispatch(getAllArtists());
      //GET ALL GENRES IN DB AND STORE THEM FOR THE BROWSE BY GENRE COMPONENT
      dispatch(getGenres());
    }

    getSeedData();
  }, [dispatch]);

  //SCROLL DOWN TO CATEGORY DIV WHEN USER SELECTS A CATEGORY
  useEffect(() => {

    function scrollToCategory() {

      if (category[0] === "Artists") {
        artistResultsRef.current.scrollIntoView({
          behavior: "smooth",
        });
      } else if (category[0] === "Genre") {
        genreResultsRef.current.scrollIntoView({
          behavior: "smooth",
        });
      } else if (category[0] === "Danceability") {
         danceabilityResultsRef.current.scrollIntoView({
          behavior: "smooth",
        });
      }

    }
    scrollToCategory();
  }, [category, setCategory]);

  //LISTENS FOR ANY CHANGES IN ERRORS IN STATE AND WILL TRIGGER FLASH MESSAGES ACCORDINGLY
  useEffect(() => {
    const displayFlashMessage = () => {

        if (lyricsError) {
          setNoLyricsFlashMessage(true);
          console.log("There is a lyrics error");
          dispatch(resetLyricsError());
        }
        if (languageError) {
          setLanguageNotFoundFlashMessage(true);
          console.log("There is a language error");
          dispatch(resetLanguageError());
        }
        if (translationError) {
          setTranslationErrorFlashMessage(true);
          console.log("Here is what translation error is: ", translationError);
          dispatch(resetTranslationError());
        }

    }
    displayFlashMessage();
  }, [languageError, translationError, lyricsError, dispatch])

////////////////////////////////////////////////////  HANDLE CLICK FUNCTIONS  ////////////////////////////////////////////////////

const handleCategoryClick = (category) => {
  setCategory([category, {}]);
  dispatch(resetStore("artists", "albums", "tracks", "lyrics", "translation"));
}

const handleNoAlbumsError = () => {
   setNoAlbumsFlashMessage(true);
}

////////////////////////////////////////////////////  JSX VARIABLES  ////////////////////////////////////////////////////

  //DISPLAY BROWSE BY ARTISTS COMPONENTS
  let BrowseByArtistsDiv;

  if (category[0] === "Artists") BrowseByArtistsDiv = (
      <div ref={artistResultsRef}>
        <BrowseByArtists handleNoAlbumsError={() => handleNoAlbumsError()} />
      </div>
  );

  //DISPLAY BROWSE BY GENRE COMPONENTS
  let BrowseByGenreDiv;

  if (category[0] === "Genre") BrowseByGenreDiv = (
      <div ref={genreResultsRef}>
        <BrowseByGenre />
      </div>
  );

  //DISPLAY BROWSE BY DANCEABILITY COMPONENTS
  let BrowseByDanceabilityDiv;

  if (category[0] === "Danceability") BrowseByDanceabilityDiv = (
      <div ref={danceabilityResultsRef}>
        <BrowseByDanceability />
      </div>
  );

////////////////////////////////////////////////////  RETURN  ////////////////////////////////////////////////////

  return (
    <div className="Browse">
      <div className="Flash-Messages-Container">
        {searchFlashMessage && (<FlashMessage setState={setSearchFlashMessage} message="We couldn't find any songs with that Artist or Song name, please try again."/> )}
        {noLyricsFlashMessage && (<FlashMessage setState={setNoLyricsFlashMessage} message="Unfortunately there are no Lyrics for that song yet."/> )}
        {languageNotFoundFlashMessage && (<FlashMessage setState={setLanguageNotFoundFlashMessage} message="That Language was not found, please try again."/> )}
        {translationErrorFlashMessage && (<FlashMessage setState={setTranslationErrorFlashMessage} message="Sorry, we couldn't get a translation at this moment."/> )}
        {noAlbumsFlashMessage && (<FlashMessage setState={setNoAlbumsFlashMessage} message="Sorry, there are no albums for that artist at this time."/> )}
      </div>
      <div className="Browse-Landing">
        <button onClick={() => handleCategoryClick("Artists")}>Artists</button>
        <button onClick={() => handleCategoryClick("Genre")}>Genre</button>
        <button onClick={() => handleCategoryClick("Danceability")}>Danceability</button>
      </div>
      {BrowseByArtistsDiv}
      {BrowseByGenreDiv}
      {BrowseByDanceabilityDiv}
    </div>
  );
}

export default Browse;
