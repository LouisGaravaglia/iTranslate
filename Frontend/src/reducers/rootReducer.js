import {combineReducers} from "redux";
import translationReducer from "./translationReducer";
import errorsReducer from "./errorsReducer";
import lyricsReducer from "./lyricsReducer";
import resultsReducer from "./resultsReducer";

const rootReducer = combineReducers({results: resultsReducer, translation: translationReducer, lyrics: lyricsReducer, errors: errorsReducer});

export default rootReducer;