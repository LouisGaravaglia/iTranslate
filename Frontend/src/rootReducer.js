// const SEED_ARTISTS = [{name: "J Balvin", id:"1vyhD5VmyZ7KMfW5gqLgo5"}, {name: "Bad Bunny", id:"4q3ewBCX7sLwd24euuV69X"}, {name: "Maluma", id:"1r4hJ1h58CWwUQe3MxPuau"}, {name: "Rosal√≠a", id:"7ltDVBr6mKbRvohxheJ9h1"}];
// const INITIAL_STATE = {songs:[], artists: SEED_ARTISTS, albums:[]};
import {GET_TRANSLATION} from "./actionTypes";
const ERRORS = {searchError: false, lyricsError: false, translationError: false}
const INITIAL_STATE = {translation: "", errors: ERRORS }

function rootReducer(state=INITIAL_STATE, action) {
    switch(action.type) {
        case GET_TRANSLATION:
        
            // const updatedStateWithErrors = {...state, state.errors.lyricsError = action.errors.lyricsError}
        return {...state, translation: action.translation}

        default:
            return state;
    }
}

export default rootReducer;
