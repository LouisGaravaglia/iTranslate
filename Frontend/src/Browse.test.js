import React from 'react'
import {render, fireEvent, waitFor, screen, cleanup} from '@testing-library/react'
import {MemoryRouter} from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect'
import { Provider } from 'react-redux'
import { createStore } from "redux";
import rootReducer from "./reducers/rootReducer";
import Browse from "./Browse";

afterEach(cleanup);

const startingState = {};

//REDUCER FUNCTION TO MIMIC REDUX REDUCER FOR STORE
function reducer(state = startingState, action) {
  switch (action.type) {
    case "ADD_ARTISTS":
      return {...state, artists: action.artists};
    default:
      return state;
  };
};

//FUNCTION TO ALLOW THE COMPONENT TO BE RENDERED USING OUR MAKESHIF REDUX STORE
function renderWithRedux(
  component,
  {initialState, store = createStore(reducer, initialState)} = {}
) {
  return {
    ...render(<Provider store={store}>{component}</Provider>)
  };
};

//SMOKE TEST
describe('Smoke Test for Browse component', () => {

  it('renders without crashing', () => {
    renderWithRedux(<Browse />)
  });

});

//SNAPSHOT TEST
describe('Snapshot Test for Browse component', () => {

  it('matches snapshot', () => {
    const {asFragment} = renderWithRedux(<Browse />)
    expect(asFragment()).toMatchSnapshot();
  });

});
