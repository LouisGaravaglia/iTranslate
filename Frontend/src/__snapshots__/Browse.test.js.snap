// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test for Browse component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      style="background-color: rgb(135, 0, 176);"
    >
      <div
        style="opacity: 1; margin-left: 2500px;"
      >
        <div
          class="Browse-Landing"
        >
          <p>
            Artists
          </p>
          <p>
            Genre
          </p>
          <p>
            Danceability
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
